zcash_primitives/src/transaction/mod.rs

```diff
@@ -576,6 +881,98 @@ impl Transaction {
         Ok(())
     }
 
+    pub fn write_v5<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.write_v5_header(&mut writer)?;
+        self.write_transparent(&mut writer)?;
+        self.write_v5_sapling(&mut writer)?;
+        self.write_v5_orchard(&mut writer)?;
+        #[cfg(feature = "zfuture")]
+        self.write_tze(&mut writer)?;
+        Ok(())
+    }
+
+    pub fn write_v5_header<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.version.write(&mut writer)?;
+        writer.write_u32::<LittleEndian>(u32::from(self.consensus_branch_id))?;
+        writer.write_u32::<LittleEndian>(self.lock_time)?;
+        writer.write_u32::<LittleEndian>(u32::from(self.expiry_height))?;
+        Ok(())
+    }
+
+    pub fn write_v5_sapling<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        if let Some(bundle) = &self.sapling_bundle {
+            Vector::write(&mut writer, &bundle.shielded_spends, |w, e| {
+                e.write_v5_without_witness_data(w)
+            })?;
+
+            Vector::write(&mut writer, &bundle.shielded_outputs, |w, e| {
+                e.write_v5_without_proof(w)
+            })?;
+
+            if !(bundle.shielded_spends.is_empty() && bundle.shielded_outputs.is_empty()) {
+                writer.write_all(&bundle.value_balance.to_i64_le_bytes())?;
+            }
+            if !bundle.shielded_spends.is_empty() {
+                writer.write_all(bundle.shielded_spends[0].anchor.to_repr().as_ref())?;
+            }
+
+            Array::write(
+                &mut writer,
+                bundle.shielded_spends.iter().map(|s| s.zkproof),
+                |w, e| w.write_all(e),
+            )?;
+            Array::write(
+                &mut writer,
+                bundle.shielded_spends.iter().map(|s| s.spend_auth_sig),
+                |w, e| e.write(w),
+            )?;
+
+            Array::write(
+                &mut writer,
+                bundle.shielded_outputs.iter().map(|s| s.zkproof),
+                |w, e| w.write_all(e),
+            )?;
+
+            if !bundle.shielded_spends.is_empty() || !bundle.shielded_outputs.is_empty() {
```

Comment:
```suggestion
            if !(bundle.shielded_spends.is_empty() && bundle.shielded_outputs.is_empty()) {
```
For consistency with `value_balance` above.